### PR TITLE
fix(query): normalize and reject legacy sort_order values

### LINKED ISSUE #296: Unknown sort attribute surfaces as HTTP 500 instead of 400 (both SortBy and Sort surfaces)
Spun out of #240's final review (PR pending).

`buildAttributeOrders` (now in `internal/entity_query_sort.go`) returns a plain error for an unknown sort attribute:

```
cannot sort by unknown attribute '<attr>' in schema '<schema>'
```

It does not wrap `forma.ErrInvalidInput`, so `classifyManagerError` maps it to HTTP 500. A caller typo in `sort_by` — or, since #240, in the structured `sort` field — is user input validation and should surface as 4xx per `docs/error-handling.md` (write-path validation wraps `ErrInvalidInput`).

#240 deliberately kept the new `Sort` path byte-compatible with the legacy message/classification so the two surfaces don't fork their contract (`TestBuildAttributeOrdersUnknownSortAttribute` pins the exact message). Fix should reclassify both surfaces in one coordinated change:

- wrap the unknown-attribute error with `forma.ErrInvalidInput` (keep the message text; keep naming attr + schema)
- update the pin test to assert `errors.Is(err, forma.ErrInvalidInput)` alongside the message
- audit for any operator dashboard/alerting that keys on the current 500

Related pre-existing gap noted in #240's PR body: legacy JSON-body `sort_order` values other than lowercase `asc`/`desc` are silently treated as non-desc (no validation) — worth deciding in the same pass whether to normalize or reject.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by iceboundrock**
Closes #296 (remaining scope; the unknown-attribute classification half was completed in #307).

## What

The legacy JSON-body `sort_order` field now goes through the same `normalizeSortOrder` as structured `sort[].sort_order` entries: case-insensitive fold to asc/desc, anything else rejected with a 400 `forma.ErrInvalidInput` carrier. Previously it was passed through verbatim and everything that was not exactly lowercase `"desc"` silently sorted ascending (`AttributeOrder.IsDescending()` compares `== "desc"`).

## Behavior changes (legacy JSON-body surface only)

- `"DESC"` / `"Desc"`: previously silently ascending → now descending (the direction the caller asked for).
- `"descending"` and other garbage: previously silently ascending → now HTTP 400, body `invalid sort_order '<value>': expected 'asc' or 'desc'`.
- Garbage `sort_order` with empty `sort_by`: previously silently ignored → now 400 (non-empty `SortOrder` is always validated; matches the HTTP query-param layer's existing "sort_order requires sort_by" stance).
- `""` and lowercase `"asc"`/`"desc"`: unchanged.

The query-param surface (`parseSortParams`) and structured `Sort` surface already behaved this way; this unifies the third surface on the same contract. Design discussion and approved spec: see #296 comments.

Also carries the AGENTS.md "Specs and Design Docs" convention (planning artifacts are posted to their GitHub issue, not committed — `docs/superpowers/` is gitignored) as a separate docs commit.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by iceboundrock** [comment]
## Findings

**No blocking findings.** The change matches the approved design and is behaviorally correct. Two minor, non-blocking items:

1. **Minor — legacy invalid-direction test doesn't pin the *published* message.** `TestBuildAttributeOrdersLegacyInvalidDirectionMessage` asserts `strings.Contains(err.Error(), want)`. `err.Error()` is the operator copy; since #313 the HTTP boundary publishes only `ResolvePublicMessage` output. A publication-chain regression (e.g. the error no longer carrying `ErrInvalidInput`, or the wrap becoming a bare sentinel wrap) would fail `errors.Is` today — good — but a wrong/missing published body would still pass. The structured surface pins the published body via `publicMessageOf` in `TestSortErrorsPublishTheirMessages`; the legacy surface should do the same, e.g. `want := "invalid sort_order 'descending': expected 'asc' or 'desc'"` and `publicMessageOf(t, err) == want`.

2. **Minor — new boundary contract is undocumented.** The empty-`sort_by` + garbage-`sort_order` → 400 case (previously silently ignored) is the most surprising behavior change and is covered by a test, but `docs/cn/api.md` (line 40) only notes "case-insensitive; other values → 400". A half-line calling out "non-empty `sort_order` is validated even with empty `sort_by`" would close the doc gap.

Both are suggestions; neither blocks merge.

**Comment by iceboundrock** [comment]
## Spec compliance

Approved design is posted to #296 comments (2026-08-05): normalize-and-reject on the legacy surface; reuse the existing `normalizeSortOrder` unchanged; keep `server.go`'s query-param path (`parseSortParams`) untouched.

| Design requirement | Implementation | Status |
|---|---|---|
| Legacy branch delegates to `normalizeSortOrder` ("one throat") | `resolveSortKeys` legacy branch replaced hand-rolled empty→asc with `normalizeSortOrder(req.SortOrder)` | ✅ |
| No new functions; `normalizeSortOrder` unchanged | Verified, file diff confirms | ✅ |
| `parseSortParams` / `server.go` not touched | Not in diff | ✅ |
| `""` → asc unchanged | Covered by existing `TestBuildAttributeOrdersLegacyDefaultAsc` + `normalizeSortOrder` empty case | ✅ |
| `"DESC"`/`"Desc"` → desc (behavior change) | `TestBuildAttributeOrdersLegacyCaseInsensitiveDesc` | ✅ |
| Garbage → `InvalidInputf("invalid sort_order '<v>': expected 'asc' or 'desc'")` → 400 | `TestBuildAttributeOrdersLegacyInvalidDirectionMessage` pins carrier + prose | ✅ |
| Boundary contract: non-empty `SortOrder` always validated, even with empty `SortBy` | `legacy_direction_without_sortby` case in `TestBuildAttributeOrdersValidation` | ✅ |
| Design's three test bullets | All three present and passing | ✅ |

The docs commit (`docs/cn/api.md`) is additive beyond the spec but accurate to the implemented behavior. No deviations from the approved spec.

**Comment by iceboundrock** [comment]
## Standards

- **Error semantics** (`docs/error-handling.md`, #301/#313): correct. Leaf is built with `InvalidInputf` (write-path validation carrier); the published message names the offending value and the expected state. Classification is sentinel-based (`errors.Is(err, forma.ErrInvalidInput)` → 400), no string matching.
- **Publication shape**: the legacy error propagates through `buildAttributeOrders`'s plain `fmt.Errorf("failed to resolve sort keys for schema '%s': %w", ...)` — an operator wrap. `ResolvePublicMessage` steps over plain wraps, so the prefix stays in the log and the body is exactly `invalid sort_order '<v>': expected 'asc' or 'desc'`. Correct deny-by-default shape; `TestNoBareSentinelWraps` unaffected (no bare sentinel literal is wrapped).
- **Bare `return nil, err`** at the legacy branch (line 32) is technically against the "never bare return err" rule, but it's the right call here and mirrors the existing mutual-exclusion branch (line 41–44): a `WrapPublicf` would leak a prefix into the published body (contradicting the approved message), and a `fmt.Errorf` would double-wrap the schema context already added one level up. A one-line comment explaining why it stays bare would preempt a future "cleanup" that breaks the body contract.
- **Size/nesting**: files ≤500 lines, functions ≤100, flat structure. No issues.
- **Tests**: table-driven for the validation cases, message pinned via `strings.Contains` on the operator copy (see Findings for the publication-pinning nit). Error text assertions use `errors.Is` and prose-contains, consistent with the repo's #313 guidance.

Docs commit is process prose, not code — no coding-standard impact.

**Comment by iceboundrock** [comment]
## Assessment

**Approve** — with the two minor items in Findings left to the author's discretion.

- **Correctness**: verified end-to-end. The three surfaces now share one contract; the legacy surface converts silent wrong-direction (or silent ignore) into either the direction the caller asked for or a clear 400.
- **Scope discipline**: minimal, surgical diff (~15 code lines). Follows the "one throat" design without creeping into `parseSortParams` dedup (correctly flagged out of scope).
- **Risk**: lowest. The surface is the legacy JSON-body path only; `""` and lowercase `asc`/`desc` behavior is unchanged. The two behavior changes are the intended ones, both strictly more correct than before.
- **Test quality**: good — the boundary contract that is easiest to regress (empty `SortBy` + garbage `SortOrder`) is explicitly pinned. The publication-message pinning nit is the only gap.
- **CI**: all green including E2E smoke, which exercises the sort paths.

**Comment by iceboundrock** [comment]
## Observations

1. **Whitespace asymmetry across surfaces.** `parseSortParams` (`internal/httpapi/server.go:546`) trims `sort_order`; the JSON-body surfaces (legacy, and structured since #240) do not — `normalizeSortOrder` lowercases but never trims. So `?sort_order=%20desc%20` works while `{"sort_order": " desc "}` → 400. This is pre-existing on the structured surface; this PR extends the same no-trim behavior to legacy. A future unification could trim inside `normalizeSortOrder`; the design deliberately chose not to touch it. Not a defect for this PR.

2. **Precedence: structured `sort` + garbage `sort_order`.** With `len(req.Sort) > 0`, the mutual-exclusion check (line 41–44) fires before any direction validation, so such a request answers "sort cannot be combined with sort_by/sort_order" rather than "invalid sort_order 'garbage'". Both are 400 and both name a real problem; the higher-level contract firing first is acceptable and unchanged from before.

3. **AGENTS.md convention wording.** The new "Specs and Design Docs" section states planning artifacts "are not committed", but the repo carries committed design/reference docs (`docs/federated-query/design.md`, `docs/refactor-plan.md`, `docs/manifest-reconcile.md`, `docs/ci-cd-design.md`). The line between "planning artifact → issue comment" vs "reference design → repo" isn't stated; a future reader could mis-file. Consider one clarifying clause (e.g. approved/reference design docs stay in `docs/`, scratch planning goes to the issue). Also verified `docs/superpowers/` is gitignored (`.gitignore:79`), so the mechanical claim is accurate.

4. **Test coverage of case variants.** `TestBuildAttributeOrdersLegacyCaseInsensitiveDesc` covers `"DESC"` only; `"Desc"` is exercised implicitly by `ToLower`. The design specified `"DESC"`, so this is compliant — noting for completeness.

**Comment by iceboundrock** [comment]
## Verification

Performed at branch head `5a28a3c` (worktree `/Users/ruoshi/code/Lychee/LTBase/forma-issue-296`):

- **Full unit suite**: `GOCACHE=$PWD/.gocache GOFLAGS=-buildvcs=false go test . ./cdc ./cmd/... ./factory ./internal/...` — all pass.
- **Targeted**: `TestBuildAttributeOrders*` (incl. the 3 new legacy tests + `legacy_direction_without_sortby` boundary case) and `TestSortErrorsPublishTheirMessages` — all pass.
- **Publication resolution**: confirmed by inspection of `forma.ResolvePublicMessage` (`client_error.go`) that the legacy error resolves to the published body `invalid sort_order '<v>': expected 'asc' or 'desc'` — the `buildAttributeOrders` `fmt.Errorf` wrap is a plain wrap and is stepped over; the leaf `clientError` qualifies via `ErrInvalidInput`. This matches the PR body's claimed 400 body.
- **Sentinel guard**: `TestNoBareSentinelWraps` (`client_error_guard_test.go`) scans the module; no bare sentinel wrap introduced by this diff.
- **Docs claim**: `docs/superpowers/` confirmed gitignored (`.gitignore:79`), consistent with the AGENTS.md convention commit.
- **CI** (GitHub Actions): Lint ✅, Test ✅, Benchmark Smoke ✅, k6 Smoke ✅, E2E Smoke ✅.
- Note: local `golangci-lint` is not installed in this environment; relied on the pinned CI Lint run (v1.64.8).

**Comment by iceboundrock** [comment]
## Summary

PR #364 closes the remaining scope of #296: the legacy JSON-body `sort_order` now goes through the same `normalizeSortOrder` as the structured `sort[]` surface — case-insensitive fold to asc/desc, anything else rejected with a `forma.ErrInvalidInput` carrier (HTTP 400). Previously everything not exactly lowercase `"desc"` silently sorted ascending (`IsDescending()` compares `== "desc"`).

Implementation is a single-branch rewrite in `resolveSortKeys` (~15 lines) plus three targeted tests; two docs commits ride along (API doc note, AGENTS.md planning-artifacts convention). The change exactly matches the approved design posted to #296, all surfaces now share one contract, publication/classification behavior is correct under the #313 machinery, and all CI (incl. E2E smoke) is green.

No blockers. Two minor suggestions: (1) pin the legacy invalid-direction test to the *published* body via `publicMessageOf` rather than only `err.Error()`; (2) document the empty-`sort_by` + garbage-`sort_order` → 400 boundary in `docs/cn/api.md`.

**Comment by iceboundrock** [comment]
## Review response (round 1) — addressed in `e91c942`

**Findings 1 — legacy published-body pin.** Verified: `err.Error()` is indeed only the operator copy. Added a third subtest to `TestSortErrorsPublishTheirMessages` (`legacy shared direction publishes verbatim`) that drives the full `buildAttributeOrders` path and asserts `publicMessageOf(t, err) == "invalid sort_order 'descending': expected 'asc' or 'desc'"` — pinning both that the body is verbatim and that the operator wrap stays out of it. Placed alongside the structured-surface pins rather than in `entity_query_sort_test.go` so all publication pins for sort errors live in one table.

**Findings 2 — boundary contract doc.** `docs/cn/api.md:40` legacy row now reads "…other values → 400, even when `sort_by` is empty (#296)."

**Standards — bare `return nil, err`.** Comment added at the return site: bare on purpose because a `WrapPublicf` prefix would leak into the published body (no per-entry attribute to name) and `buildAttributeOrders` adds the operator schema context one level up.

**Observation 3 — AGENTS.md scope.** Clarifying clause added: the convention covers per-issue planning artifacts only; reference designs that document the system itself (e.g. `docs/federated-query/design.md`) stay in `docs/`.

**Observations not acted on:**

- **1 (whitespace asymmetry):** confirmed pre-existing on the structured surface since #240; this PR extends the same no-trim behavior rather than introducing a divergence. Trimming inside `normalizeSortOrder` would change the structured surface too, which is outside this issue's scope. Can file a follow-up issue if you want it tracked.
- **2 (mutual-exclusion precedence):** agreed acceptable — the higher-level contract firing first predates this PR, both answers are 400 naming a real caller bug, and the precedence is structurally pinned by `sort_conflicts_with_sortorder` plus control flow (the legacy branch is skipped whenever `len(req.Sort) > 0`).
- **4 (case variants):** as noted, `"Desc"`/mixed case shares the single `strings.ToLower` switch already pinned by `"DESC"`; a second case would add symmetry, not information.

Full `make test` and `make lint` green at `e91c942`.

**Review by iceboundrock — COMMENTED** [review]
**Standards finding (hard)**

`internal/entity_query_sort.go:33` returns the `normalizeSortOrder` error unchanged (`return nil, err`). This violates AGENTS.md:69 and coding-standard.md:22-33: errors must be wrapped with contextual information; bare error returns are prohibited. The adjacent comment expressly preserves the bare return, so this is not merely an existing-context exception.

File: internal/entity_query_sort.go [hunk 1]
```diff
@@ -18,13 +18,21 @@ type sortKey struct {
...
+		// The legacy surface shares one direction across all SortBy keys. It goes
+		// through the same normalizeSortOrder as structured Sort entries (#296):
+		// non-empty garbage is invalid input even when SortBy is empty, because a
+		// direction that cannot apply to anything is still a caller bug.
+		sortOrder, err := normalizeSortOrder(req.SortOrder)
+		if err != nil {
+			// Returned bare on purpose: a WrapPublicf prefix would leak into the
+			// published body (there is no per-entry attribute to name here), and
+			// buildAttributeOrders adds the operator schema context one level up.
+			return nil, err
```
Comment (iceboundrock) [inline internal/entity_query_sort.go]: **[P2] Preserve error context**

`normalizeSortOrder` errors are returned without a context wrapper here. This violates `AGENTS.md:69` and `coding-standard.md:22-33`, both of which prohibit bare error returns. Use a plain contextual wrap such as `fmt.Errorf("failed to normalize legacy sort_order: %w", err)`: it retains the existing `PublicError` in the chain, so the HTTP response still publishes only `invalid sort_order ...`; `forma.WrapPublicf` is not needed.

**Review by iceboundrock — COMMENTED** [review]
**Findings**

- [P2] `internal/entity_query_sort.go:35` returns `normalizeSortOrder` errors bare. Add a plain contextual `%w` wrap; an inline comment provides the safe form.

**Spec**

- The required legacy behavior is implemented: empty defaults to `asc`; case-insensitive `ASC`/`DESC` normalize; other values carry `forma.ErrInvalidInput` with the approved message; a non-empty invalid `sort_order` is rejected even without `sort_by`.
- The PR extends the approved scope with `docs/cn/api.md`, a publication-path test, and a clarification appended to the prescribed AGENTS block. These are accurate and useful, but the design said “No ... doc changes” and the plan prescribed the AGENTS text byte-for-byte. They are non-blocking scope expansion, not a behavioral conflict.

**Standards**

- One hard violation: the bare error return above conflicts with `AGENTS.md` and `coding-standard.md` mandatory error-context rule.
- File/function size limits, naming, and layering remain within the documented limits.

**Assessment**

- The one required error-context fix remains before approval. The implementation is otherwise correct and narrowly routes the legacy surface through the existing normalization throat.

**Observations**

- The added publication test correctly pins the important disclosure boundary: a plain operator wrapper must not change the client-visible carrier message.
- The API documentation accurately describes the new legacy contract, including validation when `sort_by` is empty.

**Verification**

- `git diff --check main...origin/pr-364`
- `GOCACHE=$PWD/.gocache-review GOFLAGS=-buildvcs=false go test ./internal -run 'TestBuildAttributeOrders|TestSortErrorsPublishTheirMessages' -v`
- `GOCACHE=$PWD/.gocache-all GOFLAGS=-buildvcs=false make test`
- `make lint`
- All passed locally. GitHub CI is also green: Test, Lint, E2E Smoke Tests, Benchmark Smoke, and k6 Smoke.

**Summary**

- 1 P2 standards finding; spec behavior passes. Fix the contextual error wrap, then this is ready to approve.

File: internal/entity_query_sort.go (hunk 1 above)
Comment (iceboundrock) [reply internal/entity_query_sort.go]: Fixed in `5d0be41` — `fmt.Errorf("failed to normalize legacy sort_order: %w", err)`. Verified the publication claim before changing: `ResolvePublicMessage` steps over the plain wrap, and the round-1 subtest `legacy shared direction publishes verbatim` (which drives this exact path via `publicMessageOf`) still passes — the body stays exactly `invalid sort_order 'descending': expected 'asc' or 'desc'`. The comment at the return site now guards the remaining trap in the other direction: plain wrap, not `WrapPublicf`, so the operator prefix never enters the published body. Full `make test` + `make lint` green.

**Comment by iceboundrock** [comment]
## Findings

No blocking findings.

## Spec

Passes the approved #296 normalize-and-reject design: the legacy branch delegates to the existing `normalizeSortOrder`; empty remains `asc`; case variants fold; invalid values carry `ErrInvalidInput` and retain the required published message; and a non-empty legacy `SortOrder` is validated even with empty `SortBy`. The query-param parser and `normalizeSortOrder` are unchanged.

The added plain `fmt.Errorf` layer differs from the plan's literal `return nil, err`, but it is operator-only context, preserves the carrier's public message, and conforms to the repository's mandatory error-context rule. The Chinese API documentation update was outside the initial "no doc changes" estimate, but accurately documents this public contract and is appropriate.

## Standards

Pass. The change is small and cohesive, uses the shared normalization path, preserves flat control flow, and wraps the propagated error with context. The invalid-input carrier remains deliberately published while the new prefix stays operator-only, as required by `docs/error-handling.md`. No documented coding-standard violations found.

## Assessment

Ready for approval. The behavioral change is intentional, confined to the legacy JSON-body surface, and aligned with the two existing sort surfaces.

## Observations

The implementation and the two documentation changes are isolated in separate commits, which keeps review and potential reversion straightforward. The added publication test protects against a future change accidentally exposing the operator-context prefix.

## Verification

- `git diff --check` and `git show --check` completed without whitespace errors.
- Local: `GOCACHE=$PWD/.gocache GOFLAGS=-buildvcs=false go test ./internal -run 'TestBuildAttributeOrders|TestSortErrorsPublishTheirMessages' -v` passed.
- GitHub Actions passed: Test, Lint, E2E Smoke Tests, Benchmark Smoke, and k6 Smoke.

## Summary

0 blocking findings. Spec and standards checks pass; PR #364 is ready for approval.
